metric_learn/base_metric.py

```diff
@@ -161,6 +161,68 @@ def transform(self, X):
     """
 
 
+class BilinearMixin(BaseMetricLearner, metaclass=ABCMeta):
+
+  def score_pairs(self, pairs):
+      r"""
+      Parameters
+      ----------
+      pairs : array-like, shape=(n_pairs, 2, n_features) or (n_pairs, 2)
+        3D Array of pairs to score, with each row corresponding to two points,
+        for 2D array of indices of pairs if the metric learner uses a
+        preprocessor.
+
+      Returns
+      -------
+      scores : `numpy.ndarray` of shape=(n_pairs,)
+        The learned Mahalanobis distance for every pair.
+      """
+      check_is_fitted(self, ['preprocessor_', 'components_'])
+      pairs = check_input(pairs, type_of_inputs='tuples',
+                          preprocessor=self.preprocessor_,
+                          estimator=self, tuple_size=2)
+      # Note: For bilinear order matters, dist(a,b) != dist(b,a)
+      # We always choose first pair first, then second pair
+      # (In contrast with Mahalanobis implementation)
+      # I dont know wich implementation performs better
+      return np.diagonal(np.dot(
```

Comment:
I think this should be faster as it performs all operations in vectorized form
You could verify this using `%timeit` or something similar

Reply:
@mvargas33 you should do simply this on your laptop and post here the results as a comment - there is no need to include it in a test :-)

Reply:
Just realizing that the first option is bad because it computes the similarities between all `n_pairs*n_pairs` pairs!
There should be a way to avoid this while keeping things full vectorized

Reply:
I ran a test with `timeit` computing similarity for pairs of for 1.000 samples of 10.000 features each. Got these results:
```
First method took 2.203281 s
Second method took 3.51009 s
```
Where first method is the one with `np.diagonal()`, and the second the one with `zip()`

Will stick to first one for now, will try to find a way to compute it more efficiently

Reply:
Yes, because the problem with the first option is that it takes quadratic computation and memory in the number of pairs, which is clearly suboptimal

(I think if you increase the number of pairs in your timing test, you will see that the second method becomes faster)

This should be the fastest / more efficient way to compute it:
`np.sum(np.dot(M, pairs[:, 1, :].T) * pairs[:, 0, :].T, axis=0)`

Reply:
Found a better solution [here](https://stackoverflow.com/questions/14758283/is-there-a-numpy-scipy-dot-product-calculating-only-the-diagonal-entries-of-the)

For 1e4 samples and 1e4 features these are the results:
```
First method took 38.912194710999984
Second method took 50.40531421000014
Third method took 23.76173380199998
```
Will stick to this better solution